### ISSUE TITLE
Metadata Merge III: merge with table groups around

### ISSUE THREAD

**Issue by iherman** [w3c/csvw#78]
This is a followup on issues No. #76 and #77 

My understanding of the current metadata document is that I may have table groups as well. Ie, the metadata in the file can look as

```
{
  "resources": [{
    "@id" : "table.csv",
    "schema" : {
      "columns" : [{
        "name" : "GID"
      },{
        "title" : "On Street",
        "name"  : "On street"
      },{
        "title" : "Species",
        "name"  : "Species"
      }]
  },{
    "@id" : "another table",
    "schema" : [{
       ...
    }]
  }]
}
```

In which case the merge with a simple, conceptual metadata becomes even more complex than in issue #77, and to properly define this gets us closer and closer to some sort of a "semantic" merge. A quick (but ugly) solution to the immediate problem is to require that the metadata _always_ uses the table group format, even if there is only one table, and therefore the conceptual metadata would have again a similar structure.

**Comment by gkellogg** [comment]
I think looking at a normalized form which always places specified (or inferred) metadata into a normalized form will simplify the algorithms. That would imply that metadata specified using a simple scheme form would be turned into the table form using the name of the associated input file for the table specification.

**Comment by iherman** [comment]
Yep, I was also thinking... maybe the way we define the metadata itself should be much more "merge" friendly. After all, how exactly the syntax goes is an arbitrary decision. My initial thinking was something like:

Instead of having objects with arrays and ID-s, what happens if we try to use objects everywhere, avoiding the use of arrays. So, this is how the core metadata structure would look like for a table group:

```
{
  "URI table 1" : {
    "schema" : {
      "columns" : { 
        "GID" : {
          "index" : 1,
          "name" : "GID"
        },
        "On Street" : {
          "index" : 2,
          "name" : "on-street",
          "datatype" : "string"
        }  
      }
    }
  },
  "URI table 2" : {
    ...
  }
}
```

Meaning that the "conceptual" metadata, instead of #76 would be something like:

```
{
  "URI" : {
    "schema" : {
      "GID" : {
        "index" : 1
      },
      "On Street" : {
        "index" : 2
      }
    }
  }
}
```

This alternative formulation would be based on the fact that (deep) merging of objects is relatively easily specifiable,  and we can try to live with a simple concatenation of arrays.

(This may require a slight change for the conversion algorithms that would use the key to a column for the creation of a predicate in the absence of a "name", but that is a minor issue.)

What I do not know how that would affect a JSON-LD processing. @gkellogg, help! :-)

**Comment by gkellogg** [comment]
Actually, processing this as JSON-LD would turn all values into arrays when expanding. The way to think about merging as JSON-LD might be to expand/flatten both metadata files, append the objects from the second processed metadata file to the first and then frame and compact to get back to the original form. However, JSON-LD compaction does not remove duplicates from arrays, but would return arrays with potentially duplicated entries. It would require a separate pass, or specific processing instructions, to remove duplicates.

Alternatively, turn each metadata file into RDF, and then re-create the merged JSON-LD form be re-serializing to JSON-LD and framing to obtain the nominal representation, or invoke a custom process. A consequence of this is that multiple values of a given entry will not have their order preserved.

This may indicate that we would not depend on JSON-LD algorithms, but normatively define the structure of the JSON and all processing algorithms necessary. IMO, if we could rely upon RDF for this it would be simpler (for some value of simple), but ordering of multiple values needs to be considered.

**Comment by iherman** [comment]
> On 28 Nov 2014, at 17:51 , Gregg Kellogg notifications@github.com wrote:
> 
> Actually, processing this as JSON-LD would turn all values into arrays when expanding. The way to think about merging as JSON-LD might be to expand/flatten both metadata files, append the objects from the second processed metadata file to the first and then frame and compact to get back to the original form. However, JSON-LD compaction does not remove duplicates from arrays, but would return arrays with potentially duplicated entries. It would require a separate pass, or specific processing instructions, to remove duplicates.
> 
> Alternatively, turn each metadata file into RDF, and then re-create the merged JSON-LD form be re-serializing to JSON-LD and framing to obtain the nominal representation, or invoke a custom process. A consequence of this is that multiple values of a given entry will not have their order preserved.
> 
> This may indicate that we would not depend on JSON-LD algorithms, but normatively define the structure of the JSON and all processing algorithms necessary. IMO, if we could rely upon RDF for this it would be simpler (for some value of simple), but ordering of multiple values needs to be considered.

Although I do not understand all the details of what you say, simply because I do not know all the details about, e.g., the framing algorithm of JSON-LD. But, if I put my RDF hat on, I can see that turning everything into RDF triples would help in defining things.

_But_ I do not want to go there! If the proper explanation, as well as the implementation, of the metadata handling have to go (in some way or other) through RDF, we have a real problem I'm afraid. Not only would we have a problem with the acceptance of what we produce, but also with the implementation. I have spent a certain amount of time locating JavaScript libraries to handle RDF and JSON-LD, and the few around seem to be very complex, with practically no documentation, no examples; even for somebody knowing RDF relatively well it becomes a major challenge to implement anything of the sort in JavaScript. I do not even want to think about the reaction of a lambda JavaScript user who has not even heard of RDF... (Oh, the happy Python users who can use RDFLib:-)

I think that we have to look at the metadata design to see if we can have a more 'merge' friendly approach. There must be a simpler way.

Ivan

P.S. The framing algorithm is not part of the JSON-LD standard because, as far as I remember, at the time the algorithm was not entirely done and tested, which also shows that we are not talking about something obvious...

> —
> Reply to this email directly or view it on GitHub.

---

Ivan Herman, W3C
Digital Publishing Activity Lead
Home: http://www.w3.org/People/Ivan/
mobile: +31-641044153
ORCID ID: http://orcid.org/0000-0003-0782-2704

**Comment by rufuspollock** [comment]
This would take us back to having the "resources" field or similar as per http://dataprotocols.org/tabular-data-package/ - we originally took that out on basis this was describing a single csv at a time but I'd be +1 for supporting this (multiple csv / excel support is really nice).

**Comment by gkellogg** [comment]
Resolved in PR #169 and on http://www.w3.org/2015/01/28-csvw-minutes.html.

**Comment by gkellogg** [comment]
RESOLUTION: We will resolve #43, #76, #77, #78 as suggested in https://lists.w3.org/Archives/Public/public-csv-wg/2015Jan/0050.html
